app.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+# project template
+# Copyright(C) 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and / or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Graph backup logic for the Thoth project."""
+
+import logging
+import os
+import subprocess
+from datetime import datetime
+from thoth.storages import __version__ as __storages__version__
+from thoth.storages import DatabaseBackupStore
+
+_LOGGER = logging.getLogger("thoth.graph_backup_job")
+__version__ = f"0.6.0+thoth_storage.{__storages__version__}"
+KNOWLEDGE_GRAPH_HOST = os.getenv("KNOWLEDGE_GRAPH_HOST", "localhost")
+KNOWLEDGE_GRAPH_PORT = os.getenv("KNOWLEDGE_GRAPH_PORT", "5432")
+KNOWLEDGE_GRAPH_USER = os.getenv("KNOWLEDGE_GRAPH_USER", "postgres")
+KNOWLEDGE_GRAPH_DATABASE = os.getenv("KNOWLEDGE_GRAPH_DATABASE", "postgres")
+os.environ["PGPASSWORD"] = os.getenv("KNOWLEDGE_GRAPH_PASSWORD", "postgres")
+
+
+def store_backup_to_ceph(backup_file: str) -> str:
+    """Upload the backup file to ceph."""
+    adapter = DatabaseBackupStore()
+    adapter.connect()
+    document_id = adapter.store_backup(os.path.abspath(backup_file), backup_file)
+    return document_id
+
+
+def main():
+    """Perform graph backup job."""
+    _LOGGER.debug("Debug mode is on.")
+
+    backup_file_name = f"pg_dump-{datetime.now().strftime('%s')}"
+    _LOGGER.info("The database backup would be taken into: %r.", backup_file_name)
+
+    _LOGGER.debug("Starting the subprocess to take the dump from database.")
+    popen = subprocess.Popen(
```

Comment:
BTW it might be a good idea to use `run_command` here from thoth-analyzer library [1]. It can handle errors for you.

[1] https://thoth-station.ninja/docs/developers/analyzer/thoth.analyzer.html#thoth.analyzer.command.run_command

Reply:
@fridex , i tired using the `run_command` , it gives me error as :
```
File "/home/hnalla/.local/share/virtualenvs/graph-backup-job-uMhXjQhe/lib/python3.6/site-packages/thoth/analyzer/command.py", line 108, in run_command
    raise CommandError(error_msg, command=command, is_json=is_json)
thoth.analyzer.command.CommandError: Command exited with non-zero status code (1): pg_dump: [archiver (db)] connection to database "hnalla" failed: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?
```
Its the same issue in the server as well, while running as a cronjob.
The subprocess.popen works well both the place with no issues.

Reply:
How did you run it? It looks like argument propagation was not done so pg_dump picked the current logged in user and tried to connect to localhost:

```python
run_command("pg_dump -h {KNOWLEDGE_GRAPH_HOST!r} -p {KNOWLEDGE_GRAPH_PORT!r} -U {KNOWLEDGE_GRAPH_USER!r} -d {KNOWLEDGE_GRAPH_DATABASE!r} -f {backup_file_dest!r}")
```

I think there were some errors with the underlying library we use (delegator.py) when it was used with a list of strings.

Reply:
so after you suggested changes, the `run_command` works.
I had read this article https://medium.com/poka-techblog/5-different-ways-to-backup-your-postgresql-database-using-python-3f06cea4f51, it suggested that the delegator.py is not good for pg_dump due to memory usage. What do you think about this @fridex ?

Reply:
They load the whole file into memory, which is indeed, not efficient for any larger database. We write the dump to a file that gets uploaded. I don't see any issue here as, at the end of the day, its just subprocess that is run. WDYT?

Reply:
Oh true, we dump the file to ceph, I thought delegator did something different then subprocess which eats up the memory, reading it again, shows the difference in std.out i.e the memory consumption. we won't have the issue as we use file storage. 
I will push the changes here :+1: .